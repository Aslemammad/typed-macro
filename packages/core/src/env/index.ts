export { EnvContext, Modules } from './env'
