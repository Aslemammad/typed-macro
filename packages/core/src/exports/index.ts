export { Exportable, ModularizedExportable } from './exports'
