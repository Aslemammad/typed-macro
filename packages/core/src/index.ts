export * from './env'
export * from './macro'
export * from './exports'
export * from './provider'
export * from '../types'
