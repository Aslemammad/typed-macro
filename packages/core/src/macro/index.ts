export {
  Macro,
  MacroHandler,
  MacroCall,
  isMacro,
  YieldTask,
  YieldUnit,
  SuspendableTransform,
} from './macro'
export { MacroContext, MacroTransformContext } from './context'
export { ImportOption, MacroHelper } from './helper'
export { Babel } from './babel'
export { State } from './state'
export { defineMacro, MacroBuilder } from './builder'
