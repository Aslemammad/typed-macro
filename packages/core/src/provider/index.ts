export { MacroProviderOptions } from './options'
export { MacroProviderHooks } from './hooks'
export {
  MacroProvider,
  isMacroProvider,
  InstantiatedMacroProvider,
} from './provider'
export { defineMacroProvider } from './builder'
