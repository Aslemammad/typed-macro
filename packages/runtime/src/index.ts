export * from './transformer'
export * from './normalizer'
export * from './renderer'
export * from './runtime'
export * from './loader'
export * from './filter'
export * from './hook'
