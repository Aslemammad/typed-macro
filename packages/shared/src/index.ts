export * from './utils'
export * from './version'
export * from './types'
