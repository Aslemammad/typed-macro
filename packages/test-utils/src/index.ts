export * from './transformer'
export * from './runtime'
export * from './renderer'
export * from './env'
