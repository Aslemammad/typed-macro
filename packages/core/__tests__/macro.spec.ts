import { isGeneratorFunction } from '@typed-macro/shared'
import { defineMacro, MacroBuilder, isMacro } from '../src'
import { createMacroRemove } from './macros'

describe('function defineMacro()', () => {
  it('should work', () => {
    const m = defineMacro('macro')
      .withCustomType('export type A = string')
      .withCustomType('export type B = number')
      .withSignature('(): void')
      .withHandler(({ path }) => {
        path.remove()
      })

    expect(m).toMatchSnapshot()
    expect(isGeneratorFunction(m.handler)).toBe(true)
  })

  it('should reject invalid macro name', () => {
    expect(() => defineMacro('ma cro ')).toThrowError()
  })

  it('should reject empty signature', () => {
    expect(() => defineMacro('macro').withSignature('')).toThrowError()
  })

  it('should reject no signature', () => {
    // when users use javascript
    expect(() =>
      (defineMacro('macro') as any as MacroBuilder).withHandler(({ path }) =>
        path.remove()
      )
    ).toThrowError()
  })

  it('should reject invalid handlers', () => {
    /* eslint-disable @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars */
    const invalidHandlers = [
      '', // not a function
      () => {}, // empty function
      async (_: any) => {}, // async function
      async function* (_: any) {}, // async generator function
    ]
    /* eslint-enable @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars */
    invalidHandlers.forEach((h) => {
      expect(() =>
        defineMacro('macro')
          .withSignature('(): void')
          .withHandler(h as any)
      ).toThrowError()
    })
  })

  it('should accept normal functions and generator functions', () => {
    /* eslint-disable @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars */
    const validHandlers = [(_: any) => {}, function* (_: any) {}]
    /* eslint-enable @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars */
    validHandlers.forEach((h) => {
      expect(() =>
        defineMacro('macro').withSignature('(): void').withHandler(h)
      ).not.toThrowError()
    })
  })
})

describe('function isMacro()', () => {
  it('should work', () => {
    expect(isMacro(createMacroRemove())).toBe(true)
    expect(isMacro('')).toBe(false)
    expect(isMacro({})).toBe(false)
  })
})
