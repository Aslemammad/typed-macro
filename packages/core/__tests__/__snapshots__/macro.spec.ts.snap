// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function defineMacro() should work 1`] = `
Object {
  "__is_macro": true,
  "handler": [Function],
  "name": "macro",
  "types": "export type A = string
export type B = number
export function macro(): void",
  "version": "1",
}
`;
